app/services/ems_storage_dashboard_service.rb

```diff
@@ -90,4 +95,36 @@ def heatmaps
       :title         => 'Used capacity [%]'
     }
   end
+
+  def agg_events
+    event_hash = {}
+    event_list = []
+    events = EmsEvent.where("ems_id=#{@ems_id}")
+
+    events.each do |event|
+      unless event_hash[event.physical_storage_name]
+        event_hash[event.physical_storage_name] = {:fixed => 0, :not_fixed => 0}
+      end
+      if event.event_type == "autosde_critical_alert_fixed"
+        event_hash[event.physical_storage_name][:fixed] = event_hash[event.physical_storage_name][:fixed] + 1
+      else
+        event_hash[event.physical_storage_name][:not_fixed] = event_hash[event.physical_storage_name][:not_fixed] + 1
+      end
+    end
```

Comment:
This whole block could be greatly simplified with group_by and transform_values, i'll add an example in a bit

Reply:
```
event_hash = events.group_by { |ev| ev[:physical_storage_name] }.transform_values { |events| fixed, not_fixed = events.partition { |ev| ev[:event_type] == "autosde_critical_alert_fixed" }; {:fixed => fixed.count, :not_fixed => not_fixed.count} }
```